File: frontend/src/api/extractTopics.ts
/** Trigger /api/resources/extract_topics + subscribe SSE progress. */
import { apiPost } from './client';
import type { SSEEvent } from './types';

export interface ExtractTopicsResponse {
  task_id: string;
}

/** POST 触发提炼主题，返回 task_id，前端再用 EventSource 连 /stream。 */
export const triggerExtractTopics = (selectedResourceIds: string[]) =>
  apiPost<ExtractTopicsResponse>('/api/resources/extract_topics', {
    selected_resource_ids: selectedResourceIds,
  });

/**
 * SSE 订阅 /api/resources/extract_topics/stream?task_id=...
 * - 5 阶段 progress 事件按序推送
 * - 终态：completed（成功）/ error（失败）
 * 返回 cleanup 函数。
 *
 * EventSource 原生 'error' 事件歧义：连接断开 / 后端 'event: error' 都触发。
 * 用 readyState 区分：CLOSED 且从未收到 error 事件 → 真断连；否则是后端关闭。
 * 由 ProgressOverlay 内部 useEffect cleanup 触发 es.close()，READYSTATE 变 CLOSED，
 * 浏览器仍会再 fire 一次 'error'，需跳过避免覆盖后端的 error payload。
 */
export function subscribeExtractTopicsProgress(
  taskId: string,
  onEvent: (e: SSEEvent) => void
): () => void {
  const es = new EventSource(
    `/api/resources/extract_topics/stream?task_id=${encodeURIComponent(taskId)}`
  );
  let sawTerminal = false;
  const handler = (kind: 'progress' | 'completed' | 'error') =>
    (e: MessageEvent) => {
      if (kind === 'completed' || kind === 'error') {
        sawTerminal = true;
      }
      onEvent({ event: kind, data: JSON.parse(e.data) } as SSEEvent);
    };
  es.addEventListener('progress', handler('progress'));
  es.addEventListener('completed', (e: MessageEvent) => {
    handler('completed')(e);
    es.close();
  });
  es.addEventListener('error', (e: MessageEvent) => {
    // 区分：后端真的发了 'event: error'（MessageEvent）vs 连接断开（Event，无 data）。
    if (e instanceof MessageEvent && e.data) {
      handler('error')(e);
    } else if (!sawTerminal && es.readyState !== EventSource.OPEN) {
      // 真正断连且未收到后端终态事件
      onEvent({
        event: 'error',
        data: { status: 'failed', payload: { error: 'lost connection' } },
      } as SSEEvent);
    }
    es.close();
  });
  return () => es.close();
}

File: backend/src/selflearn/gateway/routes/extract_topics.py
"""提炼主题 POST 触发 + SSE 流（Task 2/3）。"""
from __future__ import annotations

import json
from collections.abc import AsyncIterator
from uuid import uuid4

from fastapi import APIRouter, BackgroundTasks
from sse_starlette.sse import EventSourceResponse

from selflearn.agents.extract_topics import run_extract_topics
from selflearn.infra.redis_client import get_redis
from selflearn.progress.stages import Stage
from selflearn.progress.stream import progress_consume
from selflearn.schemas.extract_topics import (
    ExtractTopicsRequest,
    ExtractTopicsResponse,
)


router = APIRouter(prefix="/api/resources", tags=["extract_topics"])


@router.post("/extract_topics", response_model=ExtractTopicsResponse, status_code=202)
async def trigger(
    body: ExtractTopicsRequest, background_tasks: BackgroundTasks,
) -> ExtractTopicsResponse:
    task_id = str(uuid4())
    r = get_redis()
    # Task 3 SSE 路由会复用 stream:{task_id}；这里先 set 个 TTL 兜底标志，
    # 让 progress_publish 的 expire 不用依赖第一次写入时机。
    await r.set(f"stream:{task_id}", "running", ex=3600)
    # 走 BackgroundTasks 同进程执行（T1 ~ T3 阶段不做 worker 进程扩展）
    background_tasks.add_task(run_extract_topics, task_id, body.selected_resource_ids)
    return ExtractTopicsResponse(task_id=task_id)


# ---------------------------------------------------------------------------
# Task 3：SSE 真流（仿 profile.py:_stream_events）
# ---------------------------------------------------------------------------


async def _stream_extract_topics_events(
    task_id: str,
) -> AsyncIterator[dict[str, str]]:
    """SSE 事件生成器：从 Redis Stream 读进度，按终态判定关闭 SSE。

    注意：与 profile.py / level.py 不同——extract_topics 流水线对每个阶段
    都推 status="completed"（parse/llm/validate/write/done 各一次），因此
    **不能**把所有 completed 当作终态。我们只把以下两条视为终态：

    - status="failed" — 任意阶段的失败（最严重的，立即关闭）
    - status="completed" **且** stage 是 EXTRACT_TOPICS_DONE（流水线全部结束）

    这样前端 ProgressOverlay 能逐阶段点亮 5 个圆点（running→completed→下一个
    running…），最后看到 EXTRACT_TOPICS_DONE.completed 才关闭连接。
    """
    async for ev in progress_consume(task_id):
        data = json.dumps(
            {
                "stage": ev.stage.value,
                "status": ev.status,
                "payload": ev.payload,
            },
            ensure_ascii=False,
        )
        yield {"event": "progress", "data": data}
        # 终态判定：
        # - failed：任意阶段失败都立刻关闭
        # - completed：仅 Stage.EXTRACT_TOPICS_DONE 才算全部跑完
        is_terminal_failed = ev.status == "failed"
        is_terminal_done = (
            ev.status == "completed"
            and ev.stage == Stage.EXTRACT_TOPICS_DONE
        )
        if is_terminal_failed or is_terminal_done:
            final_payload = json.dumps(
                {"status": ev.status, "payload": ev.payload},
                ensure_ascii=False,
            )
            yield {
                "event": "completed" if ev.status == "completed" else "error",
                "data": final_payload,
            }
            return


@router.get("/extract_topics/stream")
async def stream_extract_topics(task_id: str) -> EventSourceResponse:
    """Task 3 SSE 真流：前端 ProgressOverlay 用 EventSource 连这里收 5 阶段进度。"""
    return EventSourceResponse(_stream_extract_topics_events(task_id))
